File: fullstack-fintech-app/maintenance-service/src/repositories/maintenance_fee_history_repository.rs
use sqlx::{Postgres, Transaction};
use bigdecimal::BigDecimal;
use crate::models::debt_collector::{CurrencyType, DebtStatus};

pub struct MaintenanceFeeHistoryRepository;

impl MaintenanceFeeHistoryRepository {
    pub async fn record_paid(
        tx: &mut Transaction<'_, Postgres>,
        user_id: i64,
        currency_type: &CurrencyType, 
        fee_amount: &BigDecimal,
    ) -> sqlx::Result<()> {
        sqlx::query(
            r#"
            INSERT INTO maintenance_fee_history
                (user_id, currency_type, fee_amount, status, paid_on)
            VALUES ($1, $2, $3, 'PAID', NOW())
            "#
        )
        .bind(user_id)
        .bind(currency_type) 
        .bind(fee_amount)
        .execute(&mut **tx)
        .await?;

        Ok(())
    }

    pub async fn record_overdue(
        tx: &mut Transaction<'_, Postgres>,
        user_id: i64,
        currency_type: &CurrencyType,  
        fee_amount: &BigDecimal,
        reason: &str,
    ) -> sqlx::Result<()> {
        sqlx::query(
            r#"
            INSERT INTO maintenance_fee_history
                (user_id, currency_type, fee_amount, status, reason)
            VALUES ($1, $2, $3, 'OVERDUE', $4)
            "#
        )
        .bind(user_id)
        .bind(currency_type)  
        .bind(fee_amount)
        .bind(reason)
        .execute(&mut **tx)
        .await?;

        Ok(())
    }
}

File: fullstack-fintech-app/resiliences-service/Repositories/MaintenanceFeeHistoryRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using resiliences_service.Configs;
using resiliences_service.Enums;
using resiliences_service.interfaces;
using resiliences_service.Models;

namespace resiliences_service.Repositories
{
     public class MaintenanceFeeHistoryRepository : IMaintenanceFeeHistoryRepository
    {
        private readonly AppDbContext _db;

        public MaintenanceFeeHistoryRepository(AppDbContext db)
        {
            _db = db;
        }

        public async Task RecordPaidAsync(long userId, CurrencyType currencyType, decimal feeAmount)
        {
            _db.MaintenanceFeeHistories.Add(new MaintenanceFeeHistory
            {
                UserId       = userId,
                CurrencyType = currencyType,
                FeeAmount    = feeAmount,
                Status       = DebtStatus.PAID,
                AttemptedOn  = DateTime.UtcNow,
                PaidOn       = DateTime.UtcNow
            });

            await _db.SaveChangesAsync();
        }

        public async Task RecordOverdueAsync(long userId, CurrencyType currencyType, decimal feeAmount, string reason)
        {
            _db.MaintenanceFeeHistories.Add(new MaintenanceFeeHistory
            {
                UserId       = userId,
                CurrencyType = currencyType,
                FeeAmount    = feeAmount,
                Status       = DebtStatus.OVERDUE,
                Reason       = reason,
                AttemptedOn  = DateTime.UtcNow
            });

            await _db.SaveChangesAsync();
        }
    }
}